### DISCUSSION TITLE
[Rewrites] Can't disable ssl validation when rewrite from http://localhost:3000 to https

### DISCUSSION THREAD

**Discussion by illiakovalenko** [vercel/next.js#21854] [Help]
**What version of Next.js are you using?**

10.0.5

**What version of Node.js are you using?**

10.16.0

**What browser are you using?**

Chrome

**What operating system are you using?**

Windows 10

**How are you deploying your application?**

next start

**Describe the Bug**

Setting environment variable NODE_TLS_REJECT_UNAUTHORIZED=0, NODE_EXTRA_CA_CERTS, tried to use [https://www.npmjs.com/package/win-ca](win-ca), does nothing to do with the "rewrites" configuration of next.config.js
Any workarounds in order to skip cert error? For example it can be used in dev mode, and it's not a problem that it's insecure.

**Expected Behavior**

When NODE_TLS_REJECT_UNAUTHORIZED=0, NODE_EXTRA_CA_CERTS is set, proxies should not be rejected by "self signed certificate"

**To Reproduce**

1. Set rewrites:
```
// next.config.js
module.exports = {
    // ...
    rewrites: () => [
        {
            source: '/example/1/:path*',
            destination: 'https://self-signed-url:9000/example/1/:path*',
            basePath: false,
        }
    ]
    //...
}
```

2. export NODE_TLS_REJECT_UNAUTHORIZED=0 && npx next dev
3. Visit http://localhost:3000/example/1:
```
Visit Error occurred proxying https://self-signed-url:9000/example/1 Error: unable to verify the first certificate
  at TLSSocket.onConnectSecure (_tls_wrap.js:1502:34)
  at TLSSocket.emit (events.js:314:20)
  at TLSSocket._finishInit (_tls_wrap.js:937:8)
  at TLSWrap.ssl.onhandshakedone (_tls_wrap.js:711:12) {
  code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'
}
```

**Comment by roman-may** [comment]
@illiakovalenko Did you find a solution for this problem? We have the same issue an can not switch to valid certificates.

**Comment by ardiansyaherwin** [comment]
is there any updates on this? we faced the same issue

**Comment by lumenwrites** [comment]
I also can't figure out how to fix this. Can someone help me out?

**Reply by roman-may** [reply]
We ended up adding a proxy api in the project for development with https://github.com/chimurai/http-proxy-middleware 

Add a rewrite to e.g /api/proxy/:slug and add the proxy there.
Let me know if you need more information.
Or maybe somebody found a better, less hacky solution for this.

**Reply by Waishnav** [reply]
You mean you created a new api endpoint in `next_app/src/api/` folder and proxied the main backend api to make it working?

**Comment by illiakovalenko** [comment]
@roman-may @lumenwrites @ardiansyaherwin 
You can look at our [docs](https://doc.sitecore.com/xp/en/developers/hd/210/sitecore-headless-development/walkthrough--configuring-sitecore-ca-certificates-for-node-js.html) 'Create the certificate file for use in a Sitecore on-prem instance' may be it will be helpful for you.
So, we used:
`NODE_TLS_REJECT_UNAUTHORIZED=0`
and `NODE_EXTRA_CA_CERTS` here we specified path to exported cert

**Reply by thany** [reply]
>  NODE_TLS_REJECT_UNAUTHORIZED=0

If only this worked. It doesn't.

> NODE_EXTRA_CA_CERTS 

It is not workable to have to configure this on every development pc separately. I want projects to be as much of a "pull, npm i, npm start" experience as humanly possible, and I don't think that's an unreasonable expectation.

**Reply by illiakovalenko** [reply]
@thany This approach was a workaround for us and not only `NODE_TLS_REJECT_UNAUTHORIZED` has to be specified

**Reply by thany** [reply]
If you're specifiying `NODE_EXTRA_CA_CERTS`, then `NODE_TLS_REJECT_UNAUTHORIZED` should not be neccesary, or is it?

**Comment by andrewmclagan** [comment]
I'm guessing everyone is encountering in local development.

In that case you should be generating your self signed certs using [mkcert](https://github.com/FiloSottile/mkcert). You then need to set the environment variable  `NODE_EXTRA_CA_CERTS` because Node has its own cert store and does not use the system root.

```bash
export NODE_EXTRA_CA_CERTS="$(mkcert -CAROOT)/rootCA.pem"
```

As per the mkcert [readme](https://github.com/FiloSottile/mkcert#using-the-root-with-nodejs).

**Reply by thany** [reply]
I just want a ready-set-go configuration without setup guff on every dev's machine.

**Comment by thany** [comment]
Why was this moved to discussions? This is a bug that needs a fix, not a discussion with a million workarounds.

**Comment by sream** [comment]
This got to be a joke. Facing the same issue with rewrites. Even tried adding NODE_EXTRA_CA_CERTS but to no success. Why should this be a requirement for development?

Edit: I've solved the issue by using Server Actions. That way I don't need rewrites and setting the flag NODE_TLS_REJECT_UNAUTHORIZED=0 is sufficient enough.

**Comment by jmotes** [comment]
I was able to get `NODE_EXTRA_CA_CERTS` to work if I put it inline with `npm run dev`. It doesn't get picked up when placed in a `.env` file:

```
NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/root.crt npm run dev
```

Quite an ugly solution, but we were able to verify our code worked in dev this way. Since we have a docker-compose setup for local dev I should be able to add it to the startup COMMAND for our frontend container where Next.js lives however.

Update: 

Adding it to the package.json also works. In my case, since I'm now using OrbStack, it looks like this:

```
"scripts": {
	"dev": "export NODE_EXTRA_CA_CERTS=/usr/local/share/ca-certificates/orbstack-root.crt && next dev --turbopack",
	"build": "next build",
	"start": "next start",
},
```

Now you can run `npm run dev` as usual.
